<h3>Tratamento de dados</h3>
<p>Antes de adentrarmos na fase de Análise Exploratória de Dados (EDA), temos que tratar:</p>
<ul>
    <li>Tipos de dados</li>
    <li>Padronização</li>
    <li>Missing values</li>
</ul>

In [145]:
import pandas as pd

In [146]:
df = pd.read_excel("../data/bruto/dataset_comercial.xlsx")
df.head()

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
0,1,2026-01-15 00:00:00,João Silva,Sudeste,Notebook,Eletrônicos,2.0,3500.0,7000.0,Pago
1,2,01/16/2026,Maria,sul,Monitor 24pol,Eletrônicos,1.0,800.5,800.5,Pendente
2,3,2026-01-18 00:00:00,joão silva,SUDESTE,Notebook,eletrônicos,1.0,3500.0,3500.0,Pago
3,4,2026-01-20 00:00:00,Carlos Souza,Nordeste,Teclado Mecânico,Acessórios,5.0,150.0,750.0,Pago
4,5,2026-01-22 00:00:00,Ana Lima,NaN,Mouse sem fio,Acessórios,2.0,80.0,160.0,Cancelado


<h4>1 - Verificação dos tipos dos dados</h4>

In [147]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID_Venda          5000 non-null   int64  
 1   Data              5000 non-null   object 
 2   Vendedor          4998 non-null   object 
 3   Regiao            3606 non-null   object 
 4   Produto           5000 non-null   object 
 5   Categoria         4050 non-null   object 
 6   Qtd               4854 non-null   float64
 7   Preco_Unit        5000 non-null   float64
 8   Total             4854 non-null   float64
 9   Status_Pagamento  3386 non-null   object 
dtypes: float64(3), int64(1), object(6)
memory usage: 390.8+ KB


<h3>Constatações</h3>
<p>Ao olharmos para os tipos de dados do nosso dataset, descobrimos que:</p>
<ul>
    <li><b>Data</b> está como <b>object</b> em vez de <b>datetime64</b></li>
    <li><b>Qtd</b> está como <b>float64</b> em vez de <b>int64</b></li>
</ul>

<h4>Convertendo Data para datetime64</h4>

In [148]:
df['Data'] = pd.to_datetime(df['Data'])

In [149]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID_Venda          5000 non-null   int64         
 1   Data              5000 non-null   datetime64[ns]
 2   Vendedor          4998 non-null   object        
 3   Regiao            3606 non-null   object        
 4   Produto           5000 non-null   object        
 5   Categoria         4050 non-null   object        
 6   Qtd               4854 non-null   float64       
 7   Preco_Unit        5000 non-null   float64       
 8   Total             4854 non-null   float64       
 9   Status_Pagamento  3386 non-null   object        
dtypes: datetime64[ns](1), float64(3), int64(1), object(5)
memory usage: 390.8+ KB


In [150]:
df['Data'].unique()

<DatetimeArray>
['2026-01-15 00:00:00', '2026-01-16 00:00:00', '2026-01-18 00:00:00',
 '2026-01-20 00:00:00', '2026-01-22 00:00:00', '2026-01-23 00:00:00',
 '2026-01-25 00:00:00', '2026-01-26 00:00:00', '2026-05-02 00:00:00',
 '2026-01-29 00:00:00',
 ...
 '2025-04-02 00:00:00', '2026-03-07 00:00:00', '2025-04-06 00:00:00',
 '2025-07-30 00:00:00', '2025-01-10 00:00:00', '2025-11-16 00:00:00',
 '2026-04-11 00:00:00', '2026-12-02 00:00:00', '2025-12-20 00:00:00',
 '2026-10-04 00:00:00']
Length: 616, dtype: datetime64[ns]

<h4>Convertendo Qtd para int64</h4>

<p>Antes de convertermos Qtd para int64, teremos que resolver os valores nulos</p>

In [151]:
df_qtd_null = df['Qtd'] == df['Qtd'].isna()

In [152]:
df[df_qtd_null].shape[0]

14

In [153]:
df[df_qtd_null].head(14)

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
249,248,2025-02-09,CARLOS,Sul,Headset,informatica,0.0,2042.56,0.0,NaN
506,505,2025-11-26,joão,Sul,Monitor,NaN,0.0,4758.54,0.0,pago
564,563,2025-02-02,João,NaN,Notebook,Periféricos,0.0,3394.80,0.0,Pendente
1307,1306,2025-05-10,maria,Sul,Notebook,Periféricos,0.0,3790.22,0.0,Atrasado
1359,1358,2025-01-22,Maria,sul,Headset,Moveis,0.0,1414.66,0.0,pago
2266,2265,2026-04-13,Ana,sul,Mouse,Periféricos,0.0,1803.19,0.0,NaN
2760,2759,2025-04-06,maria,SUDESTE,Headset,informatica,0.0,1868.02,0.0,Pago
3710,3709,2025-05-06,CARLOS,Sudeste,Cadeira,Periféricos,0.0,4545.62,0.0,pago
3909,3908,2025-06-29,João,NaN,Cadeira,Moveis,0.0,4202.56,0.0,NaN
4108,4107,2025-02-12,Maria,NaN,Headset,Moveis,0.0,1636.53,0.0,pago


<p>Quando filtramos para Qtd ausente, não temos o dado de <b>Total</b>, só temos o dado de <b>Preco_Unit</b>, portanto, teremos que remover essas linhas para não atrapalhar a análise.</p>
<p>Podemos remover, sem grandes perdas, pois são somente 14 registros, o que representa 0,28% de todo o nosso dataset.</p>

In [154]:
df = df.dropna(subset='Qtd')

In [155]:
df['Qtd'].isna().sum()

np.int64(0)

<p>Agora, podemos converter o tipo de dado da coluna <b>Qtd</b> para o tipo correto</p>

In [156]:
df['Qtd'] = df['Qtd'].astype('int64')

In [157]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4854 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID_Venda          4854 non-null   int64         
 1   Data              4854 non-null   datetime64[ns]
 2   Vendedor          4852 non-null   object        
 3   Regiao            3503 non-null   object        
 4   Produto           4854 non-null   object        
 5   Categoria         3935 non-null   object        
 6   Qtd               4854 non-null   int64         
 7   Preco_Unit        4854 non-null   float64       
 8   Total             4854 non-null   float64       
 9   Status_Pagamento  3292 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(2), object(5)
memory usage: 417.1+ KB


<h4>2 - Padronização</h4>

<h5>Vendedor</h5>

In [158]:
df['Vendedor'].unique()

array(['João Silva', 'Maria', 'joão silva', 'Carlos Souza', 'Ana Lima',
       'Pedro Alves', 'Ana L.', nan, 'Carlos S.', 'maria', 'João',
       ' joão ', 'CARLOS', 'Ana', 'Pedro'], dtype=object)

In [159]:
df['Vendedor'] = df['Vendedor'].replace({
    'joão silva': 'João Silva',
    'João': 'João Silva',
    ' joão ': 'João Silva',
    'maria': 'Maria',
    'Carlos S.': 'Carlos Souza',
    'CARLOS': 'Carlos Souza',
    'Ana L.': 'Ana Lima',
    'Ana': 'Ana Lima',
    'Pedro': 'Pedro Alves'
}) 

In [160]:
df['Vendedor'].unique()

array(['João Silva', 'Maria', 'Carlos Souza', 'Ana Lima', 'Pedro Alves',
       nan], dtype=object)

<h5>Região</h5>

In [161]:
df['Regiao'].unique()

array(['Sudeste', 'sul', 'SUDESTE', 'Nordeste', nan, 'Norte', 'NE', 'SP',
       'Sul', 'SUL'], dtype=object)

In [162]:
df['Regiao'] = df['Regiao'].replace({
    'SUDESTE': 'Sudeste',
    'SP': 'Sudeste',
    'sul': 'Sul',
    'SUL': 'Sul',
    'NE': 'Nordeste'
})

<p>Notamos aqui que tinhamos um registro de estado de valor <b>SP</b>, logo, podemos entender que se refere ao estado de São Paulo, que se situa no Sudeste.</p>

In [163]:
df['Regiao'].unique()

array(['Sudeste', 'Sul', 'Nordeste', nan, 'Norte'], dtype=object)

<h5>Produto</h5>

In [164]:
df['Produto'].unique()

array(['Notebook', 'Monitor 24pol', 'Teclado Mecânico', 'Mouse sem fio',
       'Cadeira Gamer', 'Webcam 1080p', 'Mesa de Escritório', 'Headset',
       'Monitor', 'Mouse', 'Cadeira', 'Teclado'], dtype=object)

In [165]:
df['Produto'] = df['Produto'].replace({
    'Monitor 24pol': 'Monitor'
})

<p>Percebemos que <b>Teclado Mecânico</b> e <b>Teclado</b>, <b>Cadeira Gamer</b> e <b>Cadeira</b>, <b>Mouse sem fio</b> e <b>Mouse</b> são produtos distintos, então optei por manter como estavam para não perdermos informação.</p>

In [166]:
df['Produto'].unique()

array(['Notebook', 'Monitor', 'Teclado Mecânico', 'Mouse sem fio',
       'Cadeira Gamer', 'Webcam 1080p', 'Mesa de Escritório', 'Headset',
       'Mouse', 'Cadeira', 'Teclado'], dtype=object)

<h5>Categoria</h5>

In [167]:
df['Categoria'].unique()

array(['Eletrônicos', 'eletrônicos', 'Acessórios', 'Móveis',
       'Eletronicos', 'Eletronica', 'acessorios', 'moveis', nan,
       'Informática', 'informatica', 'Moveis', 'Periféricos'],
      dtype=object)

In [168]:
df['Categoria'] = df['Categoria'].replace({
    'eletrônicos': 'Eletrônicos',
    'Eletronicos': 'Eletrônicos',
    'Eletronica': 'Eletrônicos',
    'acessorios': 'Informática',
    'Acessórios': 'Informática',
    'Periféricos': 'Informática',
    'moveis': 'Móveis',
    'Moveis': 'Móveis',
    'informatica': 'Informática'
})

In [169]:
df['Categoria'].unique()

array(['Eletrônicos', 'Informática', 'Móveis', nan], dtype=object)

<p>Como temos duas colunas que são correlatas <b>(Produto e Categoria)</b>, vamos ver se os produtos estão categorizados corretamente</p>

In [170]:
df['Produto'].unique()

array(['Notebook', 'Monitor', 'Teclado Mecânico', 'Mouse sem fio',
       'Cadeira Gamer', 'Webcam 1080p', 'Mesa de Escritório', 'Headset',
       'Mouse', 'Cadeira', 'Teclado'], dtype=object)

In [171]:
df['Categoria'].unique()

array(['Eletrônicos', 'Informática', 'Móveis', nan], dtype=object)

In [172]:
df[df['Produto'] == 'Notebook']['Categoria'].unique()

array(['Eletrônicos', 'Informática', nan, 'Móveis'], dtype=object)

<p>Notamos que já para o produto Notebook temos mais de uma Categoria, temos que corrigir isso.</p>
<p>Para economizarmos tempo e espaço, vamos aplicar a categorização logo para diversos produtos.</p>

In [173]:
de_para = {
    'Notebook': 'Informática',
    'Monitor': 'Eletrônicos',
    'Teclado Mecânico': 'Informática',
    'Mouse sem fio': 'Informática',
    'Cadeira Gamer': 'Móveis',
    'Webcam 1080p': 'Informática',
    'Mesa de Escritório': 'Móveis',
    'Headset': 'Informática',
    'Mouse': 'Informática',
    'Cadeira': 'Móveis',
    'Teclado': 'Informática'
}

In [174]:
df['Categoria'] = df['Produto'].map(de_para)

In [175]:
df['Categoria'].unique()

array(['Informática', 'Eletrônicos', 'Móveis'], dtype=object)

<h5>Qtd</h5>

In [176]:
df[df['Qtd'] < 0]

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
8,8,2026-01-26,Carlos Souza,Nordeste,Teclado Mecânico,Informática,-1,150.00,-150.00,Devolvido
32,32,2026-03-22,João Silva,Sudeste,Teclado Mecânico,Informática,-2,150.00,-300.00,Devolvido
49,49,2026-04-30,Maria,Sul,Webcam 1080p,Informática,-1,250.00,-250.00,Devolvido
50,49,2026-04-30,Maria,Sul,Webcam 1080p,Informática,-1,250.00,-250.00,Devolvido
67,66,2026-06-10,João Silva,Sudeste,Headset,Informática,-1,300.00,-300.00,Devolvido
86,85,2026-07-28,Carlos Souza,Nordeste,Mesa de Escritório,Móveis,-2,850.00,-1700.00,Devolvido
100,99,2026-09-02,Maria,Sul,Notebook,Informática,-1,3500.00,-3500.00,Devolvido
117,116,2026-10-15,João Silva,Sudeste,Monitor,Eletrônicos,-1,800.50,-800.50,Devolvido
132,131,2025-02-20,João Silva,Sudeste,Cadeira,Móveis,-2,4540.49,-9080.98,Pendente
169,168,2025-04-26,Pedro Alves,Sudeste,Teclado,Informática,-1,356.46,-427.75,Pago


In [177]:
df['Qtd'] = df['Qtd'].abs()

In [178]:
df['Qtd'].unique()

array([ 2,  1,  5,  3, 10,  4,  9, 18, 17, 11, 19, 14, 16,  7,  6, 15,  8,
       13, 20, 12,  0])

<h5>Preço Unitário</h5>

<p>Anteriormente, ao olhar para os preços unitários dos produtos, identifiquei um preço discrepante em relação aos demais e resolvi investigar:</p>

In [179]:
df[df['Preco_Unit'] > 10000]

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
16,16,2026-02-11,João Silva,Sudeste,Monitor,Eletrônicos,2,99999.99,199999.98,Pago


<p>Aqui fiz um filtro com um valor aleatório alto para verificar valores de preço unitário verdadeiramente altos.</p>

In [180]:
df['Preco_Unit'].describe()

count     4854.000000
mean      2525.489629
std       2031.964067
min         51.190000
25%       1219.520000
50%       2471.785000
75%       3696.005000
max      99999.990000
Name: Preco_Unit, dtype: float64

<p>Resolvi olhar a distribuição de dados desse dataset para entender se pode ser um erro ou simplesmente um outlier.</p>
<p>Acabei descobrindo que o valor máximo é cerca de <b>27 vezes maior</b> que o valor representante do terceiro quartil.</p>

In [181]:
df['Preco_Unit'].sort_values(ascending=False)

16      99999.99
2460     5962.53
3658     5950.94
1960     5936.70
4613     5870.02
          ...   
2400       57.23
1746       54.76
3941       54.11
4930       52.76
3123       51.19
Name: Preco_Unit, Length: 4854, dtype: float64

<p>Aqui, observamos que o valor máximo é quase <b>17 vezes maior</b> que o segundo valor no ordenamento.</p>

In [182]:
df.groupby('Produto')['Preco_Unit'].agg(['min','mean','max']).sort_values(by='max',ascending=False)

,min,mean,max
Produto,,,
Monitor,51.19,2659.840152,99999.99
Cadeira,52.76,2514.695981,5962.53
Notebook,59.22,2553.569724,5950.94
Mouse,54.11,2573.122956,5936.70
Teclado,57.23,2582.394098,5860.35
Headset,58.29,2468.979206,5849.95
Cadeira Gamer,1200.00,1200.000000,1200.00
Mesa de Escritório,850.00,850.000000,850.00
Webcam 1080p,250.00,250.000000,250.00


<p>Agrupando os preços unitários por produto e agregando pelos valores mínimo, máximo e média, observamos que o valor máximo do produto Monitor se distancia demais da sua média, enquanto que os demais produtos não se distanciam tanto.</p>

<p>Após realizar essas investigações à cerca do valor discrepante, temos informações o suficiente para considerá-lo como um erro.</p>
<p>Dito isso, para não atrapalhar a EDA, vou optar por excluir esse registro do dataset.</p>

In [183]:
df[df['Preco_Unit'] == 99999.99]	

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
16,16,2026-02-11,João Silva,Sudeste,Monitor,Eletrônicos,2,99999.99,199999.98,Pago


In [184]:
df = df.drop(index=16)

In [185]:
df['Preco_Unit'].describe()

count    4853.000000
mean     2505.404218
std      1473.467509
min        51.190000
25%      1219.450000
50%      2471.680000
75%      3695.090000
max      5962.530000
Name: Preco_Unit, dtype: float64

<h5>Total</h5>

In [186]:
df['Total'] = df['Preco_Unit'] * df['Qtd']

In [187]:
df[df['Total'] < 0]

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento


<h5>Status de Pagamento</h5>

In [188]:
df['Status_Pagamento'].unique()

array(['Pago', 'Pendente', 'Cancelado', 'Devolvido', 'Atrasado', nan,
       'pago'], dtype=object)

In [189]:
df['Status_Pagamento'] = df['Status_Pagamento'].replace({
    'pago': 'Pago'
})

In [190]:
df['Status_Pagamento'].unique()

array(['Pago', 'Pendente', 'Cancelado', 'Devolvido', 'Atrasado', nan],
      dtype=object)

In [191]:
df.head()

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
0,1,2026-01-15,João Silva,Sudeste,Notebook,Informática,2,3500.0,7000.0,Pago
1,2,2026-01-16,Maria,Sul,Monitor,Eletrônicos,1,800.5,800.5,Pendente
2,3,2026-01-18,João Silva,Sudeste,Notebook,Informática,1,3500.0,3500.0,Pago
3,4,2026-01-20,Carlos Souza,Nordeste,Teclado Mecânico,Informática,5,150.0,750.0,Pago
4,5,2026-01-22,Ana Lima,NaN,Mouse sem fio,Informática,2,80.0,160.0,Cancelado


<h4>3 - Tratamento de Missing values</h4>

<h5>Data</h5>

In [192]:
df['Data'].isna().sum()

np.int64(0)

<p>Não possuímos nenhum dado ausente de Data.</p>

<h5>Vendedor</h5>

In [193]:
df['Vendedor'].isna().sum()

np.int64(2)

In [194]:
df[df['Vendedor'].isna()]

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
25,25,2026-03-05,NaN,Sudeste,Monitor,Eletrônicos,1,800.5,800.5,Pendente
59,58,2026-05-20,NaN,Norte,Headset,Informática,2,300.0,600.0,Pago


In [195]:
df['Vendedor'] = df['Vendedor'].fillna("Não identificado")

<h5>Região</h5>

In [196]:
df['Regiao'].isna().sum()

np.int64(1351)

In [197]:
df[df['Regiao'].isna()]

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
4,5,2026-01-22,Ana Lima,NaN,Mouse sem fio,Informática,2,80.00,160.00,Cancelado
6,7,2026-01-25,Maria,NaN,Monitor,Eletrônicos,3,800.50,2401.50,Pago
43,43,2026-04-15,Pedro Alves,NaN,Notebook,Informática,1,3500.00,3500.00,Pago
74,73,2026-06-28,Pedro Alves,NaN,Webcam 1080p,Informática,1,250.00,250.00,Pago
109,108,2026-09-25,Pedro Alves,NaN,Monitor,Eletrônicos,1,800.50,800.50,Pago
...,...,...,...,...,...,...,...,...,...,...
4972,4971,2025-09-19,Maria,NaN,Monitor,Eletrônicos,10,1500.28,15002.80,Atrasado
4974,4973,2026-02-06,João Silva,NaN,Teclado,Informática,14,2437.83,34129.62,Pendente
4985,4984,2026-05-16,João Silva,NaN,Notebook,Informática,13,2595.82,33745.66,NaN
4991,4990,2026-02-13,Ana Lima,NaN,Headset,Informática,10,3570.22,35702.20,Pendente


<p>Como temos <b>1351 registros</b> sem a <b>Região</b>, iremos manter os registros e incluir "Não informado" para os valores ausentes.</p>

In [198]:
df['Regiao'] = df['Regiao'].fillna("Não informado")

In [199]:
df['Regiao'].unique()

array(['Sudeste', 'Sul', 'Nordeste', 'Não informado', 'Norte'],
      dtype=object)

<h5>Produto</h5>

In [200]:
df['Produto'].isna().sum()

np.int64(0)

<h5>Categoria</h5>

In [201]:
df['Categoria'].isna().sum()

np.int64(0)

<h5>Status de Pagamento</h5>

In [202]:
df['Status_Pagamento'].isna().sum()

np.int64(1562)

In [203]:
df[df['Status_Pagamento'].isna()]

,ID_Venda,Data,Vendedor,Regiao,Produto,Categoria,Qtd,Preco_Unit,Total,Status_Pagamento
123,122,2025-06-27,João Silva,Sul,Headset,Informática,18,2786.81,50162.58,NaN
124,123,2026-03-22,Carlos Souza,Sudeste,Notebook,Informática,17,5458.28,92790.76,NaN
127,126,2025-02-20,Ana Lima,Sul,Cadeira,Móveis,19,2838.37,53929.03,NaN
129,128,2026-05-20,Maria,Sul,Mouse,Informática,2,4077.11,8154.22,NaN
131,130,2025-01-09,Ana Lima,Sul,Notebook,Informática,1,3683.49,3683.49,NaN
...,...,...,...,...,...,...,...,...,...,...
4986,4985,2026-05-16,Maria,Sudeste,Mouse,Informática,9,4384.21,39457.89,NaN
4992,4991,2025-03-23,Carlos Souza,Sul,Mouse,Informática,2,2725.41,5450.82,NaN
4993,4992,2026-05-11,Pedro Alves,Sul,Notebook,Informática,6,5210.87,31265.22,NaN
4996,4995,2025-02-21,Maria,Não informado,Monitor,Eletrônicos,20,3293.49,65869.80,NaN


In [204]:
df['Status_Pagamento'].unique()

array(['Pago', 'Pendente', 'Cancelado', 'Devolvido', 'Atrasado', nan],
      dtype=object)

In [205]:
df.columns

Index(['ID_Venda', 'Data', 'Vendedor', 'Regiao', 'Produto', 'Categoria', 'Qtd',
       'Preco_Unit', 'Total', 'Status_Pagamento'],
      dtype='object')

In [206]:
df['Status_Pagamento'] = df['Status_Pagamento'].fillna("Não informado")

<h4>Exportação dos dados tratados</h4>

In [207]:
df.to_excel("../data/tratado/dataset_comercial_tratado.xlsx", index=False) 